Updating contribution guide

janosdebugs opened the pull request:
Updating the contribution guide with the current core team processes. Integrating common community questions.

cam72cam commented on CONTRIBUTING.md:
````diff
@@ -26,133 +26,148 @@ We specifically do not merge PRs **without prior issues** that:
+```
 
-<!-- /MarkdownTOC -->
+This will create the `opentofu` binary in the current working directory, which you can test by running `./tofu --version`.
````
`tofu`

Yantrio commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
 
-This document provides guidance on OpenTofu contribution recommended practices. It covers how to submit issues, how to get involved in the discussion, how to work on the code, and how to contribute code changes.
+1. Set up a Go development environment with Git.
+2. Pay attention to copyright: write the code yourself, avoid copy/paste. Disable or limit your AI coding assistant.
```
it may be worth linking to the DCO here for more information too.

Evi1Pumpkin replied:
I agree

Yantrio commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
+### Setting up your development environment
 
-All major changes to OpenTofu Core go through the public RFC process, including those proposed by the core team. Thus, if you'd like to propose such a change, please prepare an RFC, so that the community can discuss the change and everybody has a chance to voice their opinion. You're also welcome to voice your own opinion on existing RFCs! You can find them by [going to the issues view and filter ...
+You can develop OpenTofu on any platform you like. However, some tests are currently failing on Windows, so we recommend either a Linux (WSL works) or MacOS build environment. You will need to install [Go](https://golang.org/) and [Git](https://git-scm.com/) installed, and we recommend an IDE to help you with code completion and code quality warnings. (We recommend installing the Go version docum ...
```
```suggestion
You can develop OpenTofu on any platform you like. However, we recommend either a Linux (including WSL on Windows) or a MacOS build environment. You will need [Go](https://golang.org/) and [Git](https://git-scm.com/) installed, and we recommend an IDE to help you with code completion and code quality warnings. (We recommend installing the Go version documented in the [.go-version](.go-version) file.)
```

Yantrio commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
+You can develop OpenTofu on any platform you like. However, some tests are currently failing on Windows, so we recommend either a Linux (WSL works) or MacOS build environment. You will need to install [Go](https://golang.org/) and [Git](https://git-scm.com/) installed, and we recommend an IDE to help you with code completion and code quality warnings. (We recommend installing the Go version docum ...
 
-Generally, we appreciate external contributions very much and would love to work with you on them. **However, please make sure to read the [Contributing a Code Change](#contributing-a-code-change) section prior to making a contribution.**
+If you use Visual Studio Code or Goland/IntelliJ and have Docker or Podman installed, you can also use a [devcontainer](.devcontainer.json) instead of installing the tools directly on your machine. In Visual Studio Code, you can install the [Remote Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers), then reopen the project to get a prompt ...
```
```suggestion
Alternatively, if you use Visual Studio Code or Goland/IntelliJ and have Docker or Podman installed, you can also use a [devcontainer](.devcontainer.json). In Visual Studio Code, you can install the [Remote Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers), then reopen the project to get a prompt about activating the devcontainer. In Goland/Intellij, open the `.devcontainers.json` file and click the purple cube icon that appears next to the line numbers to activate the dev container. At this point you can proceed as if you were [building natively](#building-natively) on Linux.
```

Yantrio commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
 
----
+> [!TIP]
+> Replace the `GOOS` and `GOARCH` values with your target platform if you wish to cross-compile.
```
might be a good idea to link to the docs on this (https://pkg.go.dev/cmd/go#hdr-Compile_and_run_Go_program)

Yantrio commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
-- [Adding or updating dependencies](#adding-or-updating-dependencies)
-- [Acceptance Tests: Testing interactions with external services](#acceptance-tests-testing-interactions-with-external-services)
-- [Generated Code](#generated-code)
+If you have Docker or Podman installed, you can run the entire build process in a container too:
```
```suggestion
If you have Docker or a compatible alternative installed, you can run the entire build process in a container too:
```

Lots of people love tools like orbstack, cri/containerd etc, there's lots of alternatives too.

Yantrio commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
+This will create the `tofu` binary in the current working directory, which you can test by running `./tofu --version`.
 
-In order to contribute a code change, you should fork the repository, make your changes, and then submit a pull request. Crucially, all code changes should be preceded by an issue that you've been assigned to. If an issue for the change you'd like to introduce already exists, please communicate in the issue that you'd like to take ownership of it. If an issue doesn't yet exist, please create one  ...
+> [!TIP]
```
Do we need this tip in here twice?

Yantrio commented on CONTRIBUTING.md:
````diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
 ```
-Git has a built-in flag to append this line automatically:
+
+Alternatively, you can also run the `go test` command in the package you are currently working on.
````
maybe a link to the docs for more ways people can test (https://pkg.go.dev/cmd/go#hdr-Test_packages) ?

Yantrio commented on CONTRIBUTING.md:
````diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
+### Signing off your commits
+
+When you contribute code to OpenTofu, we require you to add a [Developer Certificate of Origin](https://developercertificate.org/) sign-off. Please read the DCO carefully before you proceed and only contribute code you have written yourself. Please do not add code that you have not written from scratch yourself without discussing it in the related issue first.
+
+The simplest way to add a sign-off is to use the `-s` command when you commit:
+
 ```
-~> git commit -s -m 'This is my commit message'
+git commit -s -m "My commit message"
 ```
 
-You can find more details about the DCO checker in the [DCO app repo](https://github.com/dcoapp/app).
+> [!IMPORTANT]
+> Make sure your `user.name` and `user.email` setting in Git matches your GitHub settings. This will allow the automated DCO check to pass and avoid delays when merging your PR.
 
-Additionally, please update [the changelog](CHANGELOG.md) if you're making any user-facing changes.
+> [!TIP]
+> Have you forgotten your sign-off? You can use `git rebase --signoff` to add it after the fact.
 
````
It may be worth mentioning that this works as a check on a pull request, and if it fails they can check the details of the failing check for more information about why.

Yantrio commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
-After that, you can either install the [Go](https://golang.org/) compiler locally, or use [Docker](https://www.docker.com/) to build in a container. At this time the OpenTofu development environment targets only Linux and MacOS systems. While OpenTofu itself is compatible with Windows, unfortunately the unit test suite currently contains Unix-specific assumptions around maximum path lengths, path ...
+We take copyright and intellectual property very seriously. A few quick rules should help you:
 
-If using Visual Studio Code or Goland/IntelliJ, a [devcontainer](.devcontainer.json) is included which integrates directly with Docker. In Visual Studio Code, you can install the [Remote Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) extension, then reopen the project to get a prompt about activating the devcontainer. In Goland/Intell ...
+1. You should write most, if not all of your code yourself.
+2. Do not copy code from other authors without permission. If you have permission, always add the `Co-authored-by` sign-off to your commits.
+3. When you copy/paste code from within the OpenTofu code, always make it explicit where you copied from.
+4. When you copy from external sources, always make explicit where your copy is coming from and make sure that their license and copyright attribution requirements are met.
+5. Specifically, do not copy from the Terraform repository, or any PRs filed against that repository. Doing so will immediately disqualify your PR from being merged.
 
-If not using the devcontainer (if using an incompatible IDE), you can still still build with Docker (and thus need not install any dependencies locally) by running docker commands directly. See the [Building with Docker](#building-with-docker) section.
+> [!WARNING]
+> Coding assistants based on large language models (LLMs), such as ChatGPT or GitHub Copilot, are awesome tools to help. However, in the specific case of OpenTofu the training data may include the BSL-licensed Terraform. Since the OpenTofu/Terraform codebase is very specific and LLMs don't have any other training source, they may emit copyrighted code. Please avoid using LLM-based coding assistan ...
```
@Evi1Pumpkin you've been writing some stuff up in this area, would you mind looking at this please?

janosdebugs replied:
I've slightly updated this section but would still like some feedback @Evi1Pumpkin

Evi1Pumpkin replied:
LGTM

janosdebugs replied:
Thank you!

Yantrio reviewed (changes requested):
Just a few nitpicks, overall I'm happy with it :D

marcinwyszynski commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
 
-Additionally, please update [the changelog](CHANGELOG.md) if you're making any user-facing changes.
+> [!TIP]
+> Have you forgotten your sign-off? You can use `git rebase --signoff` to add it after the fact.
```
I think you can also sign the DCO manually, without any Git shenanigans?

janosdebugs replied:
Yes, if you make your contribution on the GitHub web interface. Otherwise you'll have to do git shenanigans, unfortunately.

marcinwyszynski replied:
I'm not 100% sure. I always forget to sign off, push my code via the CLI and then just sign the DCO manually from the link on the red check.

janosdebugs replied:
Interesting, I'll have to check if that's available to external contributors.

janosdebugs replied:
Checked, unfortunately that option is only available to maintainers as an override.

marcinwyszynski replied:
Ah, OK, good to know.

marcinwyszynski commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
+We take copyright and intellectual property very seriously. A few quick rules should help you:
 
-If using Visual Studio Code or Goland/IntelliJ, a [devcontainer](.devcontainer.json) is included which integrates directly with Docker. In Visual Studio Code, you can install the [Remote Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) extension, then reopen the project to get a prompt about activating the devcontainer. In Goland/Intell ...
+1. You should write most, if not all of your code yourself.
```
Can we start with the note about LLMs, and then put this point in the context of that? Otherwise it's a bit unclear - does it mean I can't submit my friend's code? Or does it refer to not copying from the legacy repo?

janosdebugs replied:
Technically no, you shouldn't submit your friends code without their permission. This may need rephrasing.

janosdebugs replied:
Rephrased.

marcinwyszynski commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
+2. Do not copy code from other authors without permission. If you have permission, always add the `Co-authored-by` sign-off to your commits.
+3. When you copy/paste code from within the OpenTofu code, always make it explicit where you copied from.
+4. When you copy from external sources, always make explicit where your copy is coming from and make sure that their license and copyright attribution requirements are met.
+5. Specifically, do not copy from the Terraform repository, or any PRs filed against that repository. Doing so will immediately disqualify your PR from being merged.
```
> Doing so will immediately disqualify your PR from being merged.

...and prevent you from submitting further contributions.

janosdebugs replied:
Rephrased, maybe this works better?

Evi1Pumpkin commented on CONTRIBUTING.md:
````diff
@@ -26,133 +26,169 @@ We specifically do not merge PRs **without prior issues** that:
 
-```sh
-cd opentofu
-go install ./cmd/tofu
````
I wonder why you decided to change this explanation and use the `go build` instead? What was the reason?

janosdebugs replied:
Because it's not in line with the containerized description below and it's also not intuitive that this will build and install the binaries.

cam72cam commented:
Looks good to me, maybe we want to move FAQ up a bit?  It's kind of buried at the bottom?

cube2222 commented on CONTRIBUTING.md:
```diff
@@ -167,50 +204,104 @@ make license-check
+1. **Is it possible to implement this on a technical level?**<br />Sometimes, even if a feature would be extremely useful, the state of the codebase doesn't let us do it. 
+2. **Does the feature cause more technical debt?**<br />A feature request may hide a larger issue under the hood. Sometimes it is more desirable to resolve the underlying issue instead of implementing the feature in isolation.
+3. **Is there someone who would do the work?**<br />The core team doesn't have the capacity to implement everything, so for many issues community contributions are very welcome. Sometimes companies external to OpenTofu decide to dedicate developers for the development of a specific larger feature, which can also weigh in on the decision-making process.
+4. **Is there enough capacity on the core team to support a community contributor?**<br />We don't expect contributors to implement the entire feature in isolation, we actively participate with planning, reviews and writing documentation as needed.
```
I would clarify in the description here that:

"We don't expect contributors to implement the entire feature in isolation, we actively participate with planning, reviews and writing documentation as needed. Thus, core engineers need to be available in order for an issue to be tackled, even if it's being tackled by the community."

Rephrase as needed, I don't think my version reads too nice.

cube2222 commented on CONTRIBUTING.md:
```diff
@@ -167,50 +204,104 @@ make license-check
+
+### Can I become a core team member?
+
+Possibly, please look for open positions with the sponsoring companies as they hire core team members. The interview process is the same regardless of which sponsoring company you apply to. To become a core team member, you must be equally good at Go and at communication since much of our work is helping the community. Good luck!
```
I believe this should be "Possibly. Please look...".

serdardalgic replied:
It would be helpful to redirect people to a JOBS or Career page that lists the open positions from the sponsoring companies.

janosdebugs replied:
Sorry, I was out of office. Unfortunately, I don't think all member companies have a dedicated careers page. :(

cube2222 reviewed (approved):
LGTM!

serdardalgic commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,170 @@ We specifically do not merge PRs **without prior issues** that:
+You can develop OpenTofu on any platform you like. However, we recommend either a Linux (including WSL on Windows) or a MacOS build environment. You will need [Go](https://golang.org/) and [Git](https://git-scm.com/) installed, and we recommend an IDE to help you with code completion and code quality warnings. (We recommend installing the Go version documented in the [.go-version](.go-version) fi ...
 
-Generally, we appreciate external contributions very much and would love to work with you on them. **However, please make sure to read the [Contributing a Code Change](#contributing-a-code-change) section prior to making a contribution.**
+Alternatively, if you use Visual Studio Code or Goland/IntelliJ and have Docker or Podman installed, you can also use a [devcontainer](.devcontainer.json). In Visual Studio Code, you can install the [Remote Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers), then reopen the project to get a prompt about activating the devcontainer. In Gol ...
```
dev containers look interesting, I didn't try them, I'll check them. 

The dev containers are not directly related to how to debug the code, but I think the information in [DEBUGGING.md](https://github.com/opentofu/opentofu/blob/main/DEBUGGING.md) could be melded into this Contribution guide, or it can be mentioned there. I don't have a specific suggestion, feel free to skip this comment :)

serdardalgic commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,170 @@ We specifically do not merge PRs **without prior issues** that:
 ---
 
-## Core Team
+### Building OpenTofu
```
Similar gut feeling about the https://github.com/opentofu/opentofu/blob/main/BUILDING.md . It would make sense to meld it into here, or mention the other document here, and make sure that both of the documents are aligned.

serdardalgic commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,170 @@ We specifically do not merge PRs **without prior issues** that:
+---
 
-Refer to the file [`.go-version`](.go-version) to see which version of Go OpenTofu is currently built with. Other versions will often work, but if you run into any build or testing problems please try with the specific Go version indicated. You can optionally simplify the installation of multiple specific versions of Go on your system by installing [`goenv`](https://github.com/syndbg/goenv), whic ...
+## Advanced topics
```
Maybe this one deserves another document, like CONTRIBUTING-ADVANCED.md or something like that?

serdardalgic replied:
or DEVELOPER-GUIDE.md?

janosdebugs replied:
We could write a much larger docs, but my experience from [this project](https://containerssh.io/v0.5/development/devenv/) is that it will be read rarely. I think it should best be kept in one place and as short as possible.

serdardalgic commented on CONTRIBUTING.md:
````diff
@@ -26,133 +26,170 @@ We specifically do not merge PRs **without prior issues** that:
+
+Some files in the OpenTofu CLI codebase are generated. In most cases, we update these using `go generate`, which is the standard way to encapsulate code generation steps in a Go codebase.
 
 ```
````
I think it might be helpful to keep those examples for testing the packages.

serdardalgic commented on CONTRIBUTING.md:
```diff
@@ -26,133 +26,170 @@ We specifically do not merge PRs **without prior issues** that:
+
+> [!TIP]
+> You can find more information on testing in the [Go documentation](https://pkg.go.dev/cmd/go#hdr-Test_packages).
+
```
Why not keeping the examples for testing the packages here? 
```
go test ./internal/command/...
go test ./internal/addrs
```
I mentioned it in the old section of the PR.

serdardalgic commented on CONTRIBUTING.md:
```diff
@@ -167,50 +204,104 @@ make license-check
+---
 
-OpenTofu supports various [backends](https://opentofu.org/docs/language/settings/backends/configuration). We run integration test against them to ensure no side effects when using OpenTofu.
+## FAQ
```
I think this FAQ definitely needs a separate document of its own. 
How about moving this part to a separate document and linking it here? 

I think when the Contribution document gets bigger and longer, people might refrain reading it.

serdardalgic replied:
Oh, also, TOC please for FAQ's, or cascading questions to prevent doom scrolling through FAQ document 😅

janosdebugs replied:
I'm a bit split on ToCs, GitHub already has them auto-generated for markdown files, but their usability is... less than ideal.

serdardalgic commented on CONTRIBUTING.md:
```diff
@@ -167,50 +204,104 @@ make license-check
+---
 
-Some files in the OpenTofu CLI codebase are generated. In most cases, we update these using `go generate`, which is the standard way to encapsulate code generation steps in a Go codebase.
+### What do the labels mean?
```
It's a bit of a nit pick maybe, but would come in handy for navigation:
It would be helpful to hyperlink those labels, such as [accepted](https://github.com/opentofu/opentofu/issues?q=is%3Aopen+is%3Aissue+label%3Aaccepted), [pending-steering-committee-decision](https://github.com/opentofu/opentofu/issues?q=is%3Aopen+is%3Aissue+label%3Apending-steering-committee-decision) etc.

serdardalgic commented on CONTRIBUTING.md:
```diff
@@ -2,7 +2,7 @@
 Welcome and thank you for wanting to contribute! 
 
-## In a hurry? Here's the short version:
+## Get started
```
How about mentioning Weekly Updates here? Something along the lines of:

- Want to get the latest updates? Check the [WEEKLY_UPDATES.md](https://github.com/opentofu/opentofu/blob/main/WEEKLY_UPDATES.md) 

Bonus points if you folks want to commit to the weekly DEV gatherings on Wednesdays and mention it here with the calendar invite 😅 and maybe mention the TSC (which is mentioned at the end of the document).

janosdebugs replied:
Good point. If the Wednesday meetings become regular, we could add them here.

serdardalgic reviewed (commented):
Some bits and pieces and random thoughts on stuff. Please feel free to accept or reject any of them, as you see fit. 
Thanks for your effort @janosdebugs , I think it's really valuable to add more explanations and clarify things.

Evi1Pumpkin reviewed (approved):
LGTM!

janosdebugs commented:
Thanks for the feedback everyone, I've addressed every point I believe, except for the FAQ which I don't know what to do with yet, and I've also added the community meetings to the README too. @opentofu/core-engineers @serdardalgic can I ask you to take another look?

serdardalgic reviewed (approved):
From my perspective, everything seems satisfactory. While I personally prefer documents with a Table of Contents, I'm okay with agreeing to disagree on that matter. The latest version of the CONTRIBUTING.md is significantly improved compared to the previous one, and I believe this PR contributes positively to the project. Therefore, LGTM 👍

Evi1Pumpkin commented on CONTRIBUTING.md:
```diff
@@ -15,6 +15,10 @@ Welcome and thank you for wanting to contribute!
+- Want to know what's going on? Read the [weekly updates ➡️](WEEKLY_UPDATES.md), the [TSC summary ➡️](TSC_SUMMARY.md) or join the [community meetings ➡️](https://meet.google.com/xfm-cgms-has) on Wednesdays at 14:30 CET / 8:30 AM Eastern / 5:30 AM Western / 19:00 India time on this link: https://meet.google.com/xfm-cgms-has ([📅 calendar link](https://calendar.google.com/calendar/event?eid=NDg0aWl2 ...
+
+> [!TIP]
+> For more OpenTofu events, subscribe to the [OpenTofu Events Calendar](https://calendar.google.com/calendar/embed?src=c_3f2dd3c1fe0ef4e93ef3fc456ca2dd219a2e8fd85f6b40750af16c3df370bf93%40group.calendar.google.com)!
```
🥳

Evi1Pumpkin reviewed (approved):
Looking good with the new changes.
Because we merged the `BUILDING.md` and `DEBUGGING.md` into the `CONTRIBUTING.md`, at first, I wondered if we wanted to add a table of contents at the beginning of the file. But after some thinking, I believe the GitHub `outline` feature is sufficient if somebody wants to jump to how to debug/build tofu.

Evi1Pumpkin commented:
@serdardalgic I referred to this button in the UI:
<img width="1396" alt="Screenshot 2024-02-14 at 13 05 08" src="https://github.com/opentofu/opentofu/assets/20083583/24fdae67-c46a-45bb-8183-da728ac845d1">

serdardalgic commented:
Amazing! 🥳  I didn't know that, thanks for sharing @Evi1Pumpkin 🙇

janosdebugs commented on CONTRIBUTING.md:
```diff
@@ -167,50 +301,142 @@ make license-check
+- [`bug`](https://github.com/opentofu/opentofu/labels/bug): it's broken and needs to be fixed.
+- [`enhancement`](https://github.com/opentofu/opentofu/labels/enhancement): it's a short-form feature request. If the implementation path is unclear, an `rfc` may be needed in addition.
+- [`documentation`](https://github.com/opentofu/opentofu/labels/documentation): something that needs a description on the OpenTofu website.
+- [`rfc`](https://github.com/opentofu/opentofu/labels/rfc): a long-form feature request with details on how exactly it should be implemented.
```
I'll try to get all issue labels updated once this is merged.
